plugin/group_replication/src/certifier.cc

```diff
@@ -1191,23 +1196,42 @@ void Certifier::garbage_collect() {
     precedes them), then "t" is stable and can be removed from
     the certification info.
   */
+
+  /* get the starttime in 100ns unit */
+  ulonglong starttime = my_getsystime();
+
+  mysql_mutex_lock(&LOCK_certification_info);
+
   Certification_info::iterator it = certification_info.begin();
-  stable_gtid_set_lock->wrlock();
-  while (it != certification_info.end()) {
-    if (it->second->is_subset_not_equals(stable_gtid_set)) {
-      if (it->second->unlink() == 0) delete it->second;
-      certification_info.erase(it++);
-    } else
-      ++it;
-  }
-  stable_gtid_set_lock->unlock();
 
-  /*
-    We need to update parallel applier indexes since we do not know
-    what write sets were purged, which may cause transactions
-    last committed to be incorrectly computed.
-  */
-  increment_parallel_applier_sequence_number(true);
+  while (1) {
```

Comment:
@dutow Wouldn't this start the garbage collection process start from beginning? With the current logic, it starts from where it had left?

Reply:
No, this would only move the end condition.

Currently we have an "infinite" where loop, with a break inside an if at line 1251, if `it == certification_info.end()`

All I am saying  is that we could move that condition inside the condition of the while loop. It wouldn't change how the loop is executed, because we also have a similar condition at the start of the for loop at line 1214.

Reply:
Ok. Got it. Thanks for the clarification.